### ISSUE TITLE
LLVM 18: error: always_inline function '_mm512_loadu_ps' requires target feature 'evex512', but would be inlined into function 'LoadU' that is compiled without support for 'evex512'

### ISSUE THREAD

**Issue by LinuxUserGD** [google/highway#1864]
cb933022ba6f4dcaf5cd1d67b0627f84095f214a (master branch)
OS: Gentoo Linux amd64 musl/llvm
LLVM/Clang version: [18.0.0_pre20231104](https://github.com/llvm/llvm-project/commit/83888a5404d46667647fe36d6fa510d075ed57ea) (not reproducible with LLVM 17)

Compiling highway fails with the following errors:
```bash
[1/44] /usr/lib/ccache/bin/clang++ -DHWY_SHARED_DEFINE -DTOOLCHAIN_MISS_ASM_HWCAP_H -I/var/tmp/portage/dev-cpp/highway-9999/work/highway-9999  -O3 -pipe -march=native -mtune=native -D_FORTIFY_SOURCE=3 -g0 -stdlib=libc++ -std=c++17 -fPIE -fvisibility=hidden -fvisibility-inlines-hidden -Wno-builtin-macro-redefined -D__DATE__=\"redacted\" -D__TIMESTAMP__=\"redacted\" -D__TIME__=\"redacted\" -fmerge-all-constants -Wall -Wextra -Wconversion -Wsign-conversion -Wvla -Wnon-virtual-dtor -Wfloat-overflow-conversion -Wfloat-zero-conversion -Wfor-loop-analysis -Wgnu-redeclared-enum -Winfinite-recursion -Wself-assign -Wstring-conversion -Wtautological-overlap-compare -Wthread-safety-analysis -Wundefined-func-template -fno-cxx-exceptions -fno-slp-vectorize -fno-vectorize -fdiagnostics-show-option -fcolor-diagnostics -Wc++2a-extensions -fmath-errno -fno-exceptions -MD -MT CMakeFiles/hwy_benchmark.dir/hwy/examples/benchmark.cc.o -MF CMakeFiles/hwy_benchmark.dir/hwy/examples/benchmark.cc.o.d -o CMakeFiles/hwy_benchmark.dir/hwy/examples/benchmark.cc.o -c /var/tmp/portage/dev-cpp/highway-9999/work/highway-9999/hwy/examples/benchmark.cc
samu: job failed with status 1: /usr/lib/ccache/bin/clang++ -DHWY_SHARED_DEFINE -DTOOLCHAIN_MISS_ASM_HWCAP_H -I/var/tmp/portage/dev-cpp/highway-9999/work/highway-9999  -O3 -pipe -march=native -mtune=native -D_FORTIFY_SOURCE=3 -g0 -stdlib=libc++ -std=c++17 -fPIE -fvisibility=hidden -fvisibility-inlines-hidden -Wno-builtin-macro-redefined -D__DATE__=\"redacted\" -D__TIMESTAMP__=\"redacted\" -D__TIME__=\"redacted\" -fmerge-all-constants -Wall -Wextra -Wconversion -Wsign-conversion -Wvla -Wnon-virtual-dtor -Wfloat-overflow-conversion -Wfloat-zero-conversion -Wfor-loop-analysis -Wgnu-redeclared-enum -Winfinite-recursion -Wself-assign -Wstring-conversion -Wtautological-overlap-compare -Wthread-safety-analysis -Wundefined-func-template -fno-cxx-exceptions -fno-slp-vectorize -fno-vectorize -fdiagnostics-show-option -fcolor-diagnostics -Wc++2a-extensions -fmath-errno -fno-exceptions -MD -MT CMakeFiles/hwy_benchmark.dir/hwy/examples/benchmark.cc.o -MF CMakeFiles/hwy_benchmark.dir/hwy/examples/benchmark.cc.o.d -o CMakeFiles/hwy_benchmark.dir/hwy/examples/benchmark.cc.o -c /var/tmp/portage/dev-cpp/highway-9999/work/highway-9999/hwy/examples/benchmark.cc
In file included from /var/tmp/portage/dev-cpp/highway-9999/work/highway-9999/hwy/examples/benchmark.cc:25:
In file included from /var/tmp/portage/dev-cpp/highway-9999/work/highway-9999/hwy/foreach_target.h:105:
In file included from /var/tmp/portage/dev-cpp/highway-9999/work/highway-9999/hwy/examples/benchmark.cc:29:
In file included from /var/tmp/portage/dev-cpp/highway-9999/work/highway-9999/hwy/highway.h:410:
/var/tmp/portage/dev-cpp/highway-9999/work/highway-9999/hwy/ops/x86_512-inl.h:2962:24: error: always_inline function '_mm512_loadu_ps' requires target feature 'evex512', but would be inlined into function 'LoadU' that is compiled without support for 'evex512'
 2962 |   return Vec512<float>{_mm512_loadu_ps(p)};
      |                        ^
/var/tmp/portage/dev-cpp/highway-9999/work/highway-9999/hwy/ops/x86_512-inl.h:2962:24: error: AVX vector return of type '__m512' (vector of 16 'float' values) without 'evex512' enabled changes the ABI
/var/tmp/portage/dev-cpp/highway-9999/work/highway-9999/hwy/ops/x86_512-inl.h:1780:24: error: always_inline function '_mm512_mul_ps' requires target feature 'evex512', but would be inlined into function 'operator*' that is compiled without support for 'evex512'
 1780 |   return Vec512<float>{_mm512_mul_ps(a.raw, b.raw)};
      |                        ^
/var/tmp/portage/dev-cpp/highway-9999/work/highway-9999/hwy/ops/x86_512-inl.h:1780:24: error: AVX vector argument of type '__m512' (vector of 16 'float' values) without 'evex512' enabled changes the ABI
/var/tmp/portage/dev-cpp/highway-9999/work/highway-9999/hwy/ops/x86_512-inl.h:1078:24: error: always_inline function '_mm512_add_ps' requires target feature 'evex512', but would be inlined into function 'operator+' that is compiled without support for 'evex512'
 1078 |   return Vec512<float>{_mm512_add_ps(a.raw, b.raw)};
      |                        ^
/var/tmp/portage/dev-cpp/highway-9999/work/highway-9999/hwy/ops/x86_512-inl.h:1078:24: error: AVX vector argument of type '__m512' (vector of 16 'float' values) without 'evex512' enabled changes the ABI
/var/tmp/portage/dev-cpp/highway-9999/work/highway-9999/hwy/ops/x86_512-inl.h:3113:3: error: always_inline function '_mm512_storeu_ps' requires target feature 'evex512', but would be inlined into function 'StoreU' that is compiled without support for 'evex512'
 3113 |   _mm512_storeu_ps(p, v.raw);
      |   ^
/var/tmp/portage/dev-cpp/highway-9999/work/highway-9999/hwy/ops/x86_512-inl.h:3113:3: error: AVX vector argument of type '__m512' (vector of 16 'float' values) without 'evex512' enabled changes the ABI
/var/tmp/portage/dev-cpp/highway-9999/work/highway-9999/hwy/ops/x86_512-inl.h:306:24: error: always_inline function '_mm512_setzero_ps' requires target feature 'evex512', but would be inlined into function 'Zero' that is compiled without support for 'evex512'
  306 |   return Vec512<float>{_mm512_setzero_ps()};
      |                        ^
/var/tmp/portage/dev-cpp/highway-9999/work/highway-9999/hwy/ops/x86_512-inl.h:306:24: error: AVX vector return of type '__m512' (vector of 16 'float' values) without 'evex512' enabled changes the ABI
/var/tmp/portage/dev-cpp/highway-9999/work/highway-9999/hwy/ops/x86_512-inl.h:2937:24: error: always_inline function '_mm512_load_ps' requires target feature 'evex512', but would be inlined into function 'Load' that is compiled without support for 'evex512'
 2937 |   return Vec512<float>{_mm512_load_ps(aligned)};
      |                        ^
/var/tmp/portage/dev-cpp/highway-9999/work/highway-9999/hwy/ops/x86_512-inl.h:2937:24: error: AVX vector return of type '__m512' (vector of 16 'float' values) without 'evex512' enabled changes the ABI
/var/tmp/portage/dev-cpp/highway-9999/work/highway-9999/hwy/ops/x86_512-inl.h:2157:24: error: always_inline function '_mm512_fmadd_ps' requires target feature 'evex512', but would be inlined into function 'MulAdd' that is compiled without support for 'evex512'
 2157 |   return Vec512<float>{_mm512_fmadd_ps(mul.raw, x.raw, add.raw)};
      |                        ^
/var/tmp/portage/dev-cpp/highway-9999/work/highway-9999/hwy/ops/x86_512-inl.h:2157:24: error: AVX vector argument of type '__m512' (vector of 16 'float' values) without 'evex512' enabled changes the ABI
/var/tmp/portage/dev-cpp/highway-9999/work/highway-9999/hwy/ops/x86_512-inl.h:3381:20: error: always_inline function '_mm512_castps512_ps256' requires target feature 'evex512', but would be inlined into function 'LowerHalf' that is compiled without support for 'evex512'
 3381 |   return VFromD<D>{_mm512_castps512_ps256(v.raw)};
      |                    ^
/var/tmp/portage/dev-cpp/highway-9999/work/highway-9999/hwy/ops/x86_512-inl.h:3381:20: error: AVX vector argument of type '__m512' (vector of 16 'float' values) without 'evex512' enabled changes the ABI
/var/tmp/portage/dev-cpp/highway-9999/work/highway-9999/hwy/ops/x86_512-inl.h:4592:24: error: '__builtin_ia32_shuf_f32x4' needs target feature avx512f,evex512
 4592 |   return Vec512<float>{_mm512_shuffle_f32x4(v.raw, v.raw, _MM_PERM_CDAB)};
      |                        ^
/usr/lib/llvm/18/bin/../../../../lib/clang/18/include/avx512fintrin.h:6662:12: note: expanded from macro '_mm512_shuffle_f32x4'
 6662 |   ((__m512)__builtin_ia32_shuf_f32x4((__v16sf)(__m512)(A), \
      |            ^
In file included from /var/tmp/portage/dev-cpp/highway-9999/work/highway-9999/hwy/examples/benchmark.cc:25:
In file included from /var/tmp/portage/dev-cpp/highway-9999/work/highway-9999/hwy/foreach_target.h:105:
In file included from /var/tmp/portage/dev-cpp/highway-9999/work/highway-9999/hwy/examples/benchmark.cc:29:
In file included from /var/tmp/portage/dev-cpp/highway-9999/work/highway-9999/hwy/highway.h:410:
/var/tmp/portage/dev-cpp/highway-9999/work/highway-9999/hwy/ops/x86_512-inl.h:4607:20: error: '__builtin_ia32_shuf_f32x4' needs target feature avx512f,evex512
 4607 |   return VFromD<D>{_mm512_shuffle_f32x4(v.raw, v.raw, _MM_PERM_ABCD)};
      |                    ^
/usr/lib/llvm/18/bin/../../../../lib/clang/18/include/avx512fintrin.h:6662:12: note: expanded from macro '_mm512_shuffle_f32x4'
 6662 |   ((__m512)__builtin_ia32_shuf_f32x4((__v16sf)(__m512)(A), \
      |            ^
In file included from /var/tmp/portage/dev-cpp/highway-9999/work/highway-9999/hwy/examples/benchmark.cc:25:
In file included from /var/tmp/portage/dev-cpp/highway-9999/work/highway-9999/hwy/foreach_target.h:105:
In file included from /var/tmp/portage/dev-cpp/highway-9999/work/highway-9999/hwy/examples/benchmark.cc:29:
In file included from /var/tmp/portage/dev-cpp/highway-9999/work/highway-9999/hwy/highway.h:410:
/var/tmp/portage/dev-cpp/highway-9999/work/highway-9999/hwy/ops/x86_512-inl.h:3873:24: error: '__builtin_ia32_shufps512' needs target feature avx512f,evex512
 3873 |   return Vec512<float>{_mm512_shuffle_ps(v.raw, v.raw, _MM_PERM_ABCD)};
      |                        ^
/usr/lib/llvm/18/bin/../../../../lib/clang/18/include/avx512fintrin.h:6732:12: note: expanded from macro '_mm512_shuffle_ps'
 6732 |   ((__m512)__builtin_ia32_shufps512((__v16sf)(__m512)(A), \
      |            ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
20 errors generated.
```

Upstream report: https://github.com/llvm/llvm-project/issues/70002

**Comment by jan-wassenberg** [comment]
Thanks for reporting. This is due to the new AVX10 support which changes the required targets. It is unfortunate this was unable to avoid breaking user code.
I've commented in the LLVM issue. Hopefully it will be possible to work around this. In set_macros-inl.h, would you like to try inserting 

```
#ifdef __EVEX512__
#define HWY_TARGET_STR_AVX10_512 "evex512"
#else
#define HWY_TARGET_STR_AVX10_512
#endif
```

and then change the line `#define HWY_TARGET_STR_AVX3 \`
to ` #define HWY_TARGET_STR_AVX3 HWY_TARGET_STR_AVX10_512 \`?

**Comment by jan-wassenberg** [comment]
Actually I just noticed that your command line includes `-march=native -mtune=native`. We recommend building without such flags, especially when using dynamic dispatch (which Highway's tests do). I think that will solve the mismatch.

**Comment by LinuxUserGD** [comment]
> Actually I just noticed that your command line includes `-march=native -mtune=native`. We recommend building without such flags, especially when using dynamic dispatch (which Highway's tests do). I think that will solve the mismatch.

Removing `-march=native` (which invokes `-target-feature -avx512f`) or adding `-mevex512` flag fixes the build error.

**Comment by jan-wassenberg** [comment]
Thanks for confirming :) Closing.

**Comment by ThisNekoGuy** [comment]
On Clang-22 this is a hard error: removing `-march` does nothing and `-mevex512` was removed as a valid compiler option.

My system is also Gentoo but it's built on the Clang 22 toolchain, so this is a bit of an issue for me @jan-wassenberg

**Comment by jan-wassenberg** [comment]
hm, what exactly is the issue you are seeing? On Clang 22, I'd expect we define HWY_HAVE_EVEX512 = 0 (is that what you see?), and then we would not use `,evex512`.

**Comment by ThisNekoGuy** [comment]
Weird, I ran into this issue with Highway 1.3.0 but I don't have the issue with Highway 1.4.0

My system wasn't fully up to date with 1.4.0 (I updated to 1.3.0 recently), so I guess you can ignore what I said; no march is fine with 1.4.0, but it wasn't with 1.3.0

I thought I had the log on hand from the failures I had when I was trying to understand the cause last time but I'm thinking I might have left the logs in `/tmp` so they're gone; don't mind me, I guess. Sorry about that

**Comment by jan-wassenberg** [comment]
Yes, I think this issue was fixed between 1.3.0 and 1.4.0.
No worries, glad it's all good now :)
